ionc/ion_timestamp.c

```diff
@@ -916,12 +916,113 @@ BOOL _ion_timestamp_is_leap_year(int y)
     return is_leap_year;
 }
 
+iERR _ion_timestamp_to_utc(const ION_TIMESTAMP *ptime, ION_TIMESTAMP *pout) {
+    iENTER;
+
+    int8_t sign;
+    uint16_t  min_shift, hour_shift = 0;
+    BOOL needs_day = FALSE;
+
+    if (!ptime || !pout) FAILWITH(IERR_INVALID_ARG);
+
+    _ion_timestamp_initialize(pout);
+    pout->year = ptime->year;
+    pout->month = ptime->month;
+    pout->day = ptime->day;
+    pout->hours = ptime->hours;
+    pout->minutes = ptime->minutes;
+    pout->seconds = ptime->seconds;
+    pout->precision = ptime->precision;
+    pout->tz_offset = 0; // The result is in UTC.
+    decQuadCopy(&pout->fraction, &ptime->fraction);
+
+    if (!HAS_TZ_OFFSET(ptime)) SUCCEED();
+    sign = (int8_t)((ptime->tz_offset < 0) ? -1 : 1);
+    min_shift = (uint16_t)(sign * ptime->tz_offset);
+
+    if (min_shift >= 60 * 24) {
+        // Offsets must have a magnitude of less than one day.
+        FAILWITH(IERR_INVALID_TIMESTAMP);
+    }
+    if (min_shift >= 60) {
+        hour_shift = (uint16_t)(min_shift / 60);
+        min_shift = (uint16_t)(min_shift % 60);
+    }
+
+    pout->hours -= (sign * hour_shift);
+    pout->minutes -= (sign * min_shift);
+
+    if ((int16_t)pout->minutes < 0) {
+        pout->hours -= 1;
+        pout->minutes = (uint16_t)(pout->minutes + 60);
+    }
+    else if (pout->minutes > 59) {
+        pout->hours += 1;
+        pout->minutes = (uint16_t)(pout->minutes - 60);
+    }
+
+    if ((int16_t)pout->hours < 0) {
+        pout->day -= 1;
+        pout->hours = (uint16_t)(pout->hours + 24);
+    }
+    else if (pout->hours > 23) {
+        pout->day += 1;
+        pout->hours = (uint16_t)(pout->hours - 24);
+    }
+    else {
+        SUCCEED();
+    }
+
+    if (pout->day < 1) {
+        pout->month -= 1;
+        // Defer assignment of new day until year is known.
+        needs_day = TRUE;
+    }
+    else if (pout->day > JULIAN_DAY_PER_MONTH[_ion_timestamp_is_leap_year(pout->year)][pout->month - 1]) {
+        pout->month += 1;
+        pout->day = 1;
+    }
+    else {
+        SUCCEED();
+    }
+
+    if (pout->month < 1) {
+        pout->year -= 1;
+        pout->month = 11;
```

Comment:
Fixed in #57 . ion-tests coverage added in https://github.com/amzn/ion-tests/pull/28